cache_flat_mutation_reader.hh

```diff
@@ -730,9 +749,44 @@ void cache_flat_mutation_reader::copy_from_cache_to_buffer() {
         }
     }
 
-    // We add the row to the buffer even when it's full.
-    // This simplifies the code. For more info see #3139.
     if (_next_row_in_range) {
+        bool remove_row = false;
+
+        if (_read_context.tombstone_gc_state() // do not compact rows when tombstone_gc_state is not set (used in some unit tests)
+            && !_next_row.dummy()
+            && _snp->at_latest_version()
+            && _snp->at_oldest_version()) {
+            deletable_row& row = _next_row.latest_row();
```

Comment:
Note: a recent change in master removed `latest_row()`, because it had no users. If you rebase this onto master, just bring it back.